llvm/lib/CodeGen/GlobalISel/CallLowering.cpp

```diff
@@ -845,30 +901,45 @@ bool CallLowering::handleAssignments(ValueHandler &Handler,
                                      DstMPO, DstAlign, SrcMPO, SrcAlign,
                                      MemSize, VA);
         }
-        continue;
-      }
-
-      assert(!VA.needsCustom() && "custom loc should have been handled already");
-
-      if (i == 0 && !ThisReturnRegs.empty() &&
-          Handler.isIncomingArgumentHandler() &&
-          isTypeIsValidForThisReturn(ValVT)) {
+      } else if (i == 0 && !ThisReturnRegs.empty() &&
+                 Handler.isIncomingArgumentHandler() &&
+                 isTypeIsValidForThisReturn(ValVT)) {
         Handler.assignValueToReg(ArgReg, ThisReturnRegs[Part], VA);
-        continue;
-      }
-
-      if (Handler.isIncomingArgumentHandler())
+      } else if (Handler.isIncomingArgumentHandler()) {
         Handler.assignValueToReg(ArgReg, VA.getLocReg(), VA);
-      else {
+      } else {
         DelayedOutgoingRegAssignments.emplace_back([=, &Handler]() {
           Handler.assignValueToReg(ArgReg, VA.getLocReg(), VA);
         });
       }
+
+      // Finish the handling of indirect parameter passing when receiving
+      // the value (we are in the called function or the caller when receiving
+      // the return value).
+      if (VA.getLocInfo() == CCValAssign::Indirect &&
+          Handler.isIncomingArgumentHandler()) {
+        Align Alignment = DL.getABITypeAlign(Args[i].Ty);
+        MachinePointerInfo MPO(Args[i].Regs[0]);
```

Comment:
This MPO is constructed with register number reinterpreted as address space.
--
Is it possible to use Handler.getStackAddress() here? It should work for both incoming and outgoing values.

Reply:
I think it is only possible, if we have a `VA.isMemLoc()`, but we don't always have that. I will look into it more deeply.

Reply:
Sorry, I edidted the comment after you've read it.
I thought the load below is loading a pointer, but it is loading the pointee, so it probably shoulnd't call getStackAddress.
The MPO construction should still be fixed though.

Reply:
Oh and we probably should have byVal too.

Reply:
Where would we use the stack address? We could not use it for the MPO I think, because the `getStackAddress` call already requires an MPO.

Reply:
We need a correct MPO for the load below. Currently it is not correct.

> `getStackAddress` call already requires an MPO.

It doesn't. It is an output parameter passed by reference.